### ISSUE TITLE
add "greyscale when disabled" option to sun buttons

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#658]
In https://github.com/phetsims/sun/issues/644, we moved responsibility for the disabled look of sun buttons from "appearance strategies" to EnabledNode.  This standardizes the API and UX for enabledProperty across all PhET UI components.   But as noted in https://github.com/phetsims/sun/issues/644#issuecomment-715630739, it means that buttons do not change to a gray color palette when disabled.  Like all other PhET UI components, they change their opacity to indicate whether they are enabled or disabled.

Anticipating pushback on this change, @ariel-phet, @jonathanolson and I discussed providing grayscale-when-disabled as an option (perhaps with default `true`) for sun buttons. To support this, @jonathanolson is investigating adding grayscale (and other filter features) to scenery. See https://github.com/phetsims/scenery/issues/707#issuecomment-717604465.  Assuming that works out, we can add a grayscale-when-disabled option to EnabledNode, which would be turned on only for sun buttons.

If the scenery approach doesn't pan out, another "partial revert" approach is:
* restore disabled states to ButtonInteractionState and RadioButtonInteractionState
* restore disabled state handling to buttonAppearanceStrategy (and contentAppearanceStrategy?), but handle only color palette, and leave opacity to EnabledNode

This approach is much less desirable (and really not recommended) because responsibility for the look of  disabled buttons would be split between appearance strategies and EnabledNode.

**Comment by jbphet** [comment]
Discussed in the 10/29/2020 dev meeting, and we are going forward with this but need to determine what the defaults will be.

**Comment by zepumph** [comment]
I found a manual example of this being done in FAMB/js/common/view/SliderKnob.js.

**Comment by ariel-phet** [comment]
@jonathanolson I think published Gravity and Orbits shows a good example of a grayed-out strategy that works well for both a dark and light background as shown with projector mode (the "clear" button)

**Comment by ariel-phet** [comment]
For Sun buttons we will go with a default of grayscale. 

fter reviewing sims as part of regressions testing, it is clear to me that opacity was an inferior strategy in certain cases, where as grayscale worked reasonably in pretty much all cases (although I would have an aesthetic leaning towards opacity in a few circumstances, the eraser button is one that comes to mind). Several designers and developers have expressed a preference for grayscale, that look and behavior was specifically designed in the past, and it seems like the default already for most of our sims.

It will of course be nice to have the option for the opacity strategy, but it looks like the majority of cases will be grayscale.

**Comment by jonathanolson** [comment]
I've added that default, but we'll need to talk about strategies and where those should live. Because the behavior is currently in Node's `onEnabledPropertyChange`, I've overridden that in ButtonNode (however pickability shouldn't be connected, and there should be a better way of overriding and customizing).

Tagging for developer meeting to discuss.

**Comment by zepumph** [comment]
Partially designed greyscale filter:

1. greyscale
1. opacity only on content
3. some brightness?

```
Index: js/buttons/ButtonNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/buttons/ButtonNode.js	(revision 35bec9f7110db882fa02305dc099ce42cfa04f1e)
+++ js/buttons/ButtonNode.js	(date 1604601126787)
@@ -11,6 +11,7 @@
 import merge from '../../../phet-core/js/merge.js';
 import AlignBox from '../../../scenery/js/nodes/AlignBox.js';
 import Node from '../../../scenery/js/nodes/Node.js';
+import Brightness from '../../../scenery/js/util/Brightness.js';
 import Grayscale from '../../../scenery/js/util/Grayscale.js';
 import PaintColorProperty from '../../../scenery/js/util/PaintColorProperty.js';
 import Tandem from '../../../tandem/js/Tandem.js';
@@ -118,6 +119,7 @@
     let alignBox = null;
     if ( options.content ) {
 
+      this.content = options.content;
       // For performance, in case content is a complicated icon or shape.
       // See https://github.com/phetsims/sun/issues/654#issuecomment-718944669
       options.content.pickable = false;
@@ -206,8 +208,12 @@
   onEnabledPropertyChange( enabled ) {
     super.onEnabledPropertyChange( enabled );
 
-    this.opacity = 1.0;
-    this.filters = enabled ? [] : [ Grayscale.FULL ];
+    this.opacity = 1;
+    if( this.content){
+      this.content.opacity = enabled ? 1 : .45;
+    }
+    this.filters = enabled ? [] : [ Grayscale.FULL, new Brightness( .5) ];
+    // this.content.filters = [new Opacity(.30)]
   }
 }
 

```

**Comment by arouinfar** [comment]
It's probably best to iterate on this in real-time. @jonathanolson @samreid I'm available after 3 today.

**Comment by samreid** [comment]
@jessegreenberg pointed out that the ExpandCollapse button should be red/green, but it got set to blue after the commits.

We changed `baseColorProperty` to a `DerivedProperty` so setting a `paint` attribute on that isn’t having the expected behavior.  @jonathanolson can you please investigate?

**Comment by zepumph** [comment]
Introduced by https://github.com/phetsims/sun/commit/678bd58d861f114d262d8519fe0ca54f84e1cef6, and fixed above. We want baseColorProperty to respect enabled, while still being able to be mutated. I felt like an customizable implementation Property was the best solution for now.

**Comment by arouinfar** [comment]
Please re-assign to me when things are ready to review.

**Comment by zepumph** [comment]
Ready!

**Comment by zepumph** [comment]
One change that has come from https://github.com/phetsims/sun/commit/678bd58d861f114d262d8519fe0ca54f84e1cef6 is that now the base color changes when becoming disabled. In the sun demo, the previous "transparent button" used alpha in the color to achieve transparency. Knowing that, it makes sense that the disabled state of that button is fully opaque. I added an example of a transparent button using `Node.opacity` to show how the disabled states between the two are different:

![image](https://user-images.githubusercontent.com/6856943/98420290-55cf4500-2033-11eb-9a8c-cb3f1c6e39a2.png)

**Comment by jonathanolson** [comment]
@zepumph thanks for the base color fix!

**Comment by zepumph** [comment]
I also factored this enabled logic into an option, so that it can be overridden. I have done all that I will for this issue. @arouinfar please review, and then assign to @jonathanolson for a spot check if all looks good.

**Comment by arouinfar** [comment]
I reviewed the sun button demo in master, as well as a handful of random sims in studio. I'm generally happy with the appearance of disabled sun buttons. I did notice a few quirks, though I'm not sure how much of it is related to this issue specifically.

* It seems like the pickableProperty of disabled buttons remains true. The button gets brighter on hover and the cursor remains a hand. I think this behavior suggests that the disabled button is still interactive, which isn't great UX. It looks like pickable might be problematic and is up for further discussion in #666.
* The carrot-like arrow used on carousel buttons and the pickers in the Fractions suite gets clipped when the button is disabled. For example:
  * Fractions: Mixed Numbers
![image](https://user-images.githubusercontent.com/8419308/98421666-16aaef00-2047-11eb-92f5-099b6e79de77.png)
  * Balancing Act
![image](https://user-images.githubusercontent.com/8419308/98422074-021b2680-2048-11eb-81c6-503c02c6eeb6.png)
* The disabled carousel button in the above screenshot looks odd. From https://github.com/phetsims/balancing-act/issues/97#issuecomment-577802513, it seems like there shouldn't be any background on the disabled button. My guess is that this is a sim-specific issue.
* The current strategy doesn't work well for buttons with a pure white icon, such as the ExpandCollapseButton. I don't think the ExpandCollapseButton is ever disabled in a phet-brand sim. In terms of PhET-iO customization, hiding the ExpandCollapseButton button is almost always preferable to disabling it. Buttons with a light-color icon like the erase button still look fine while disabled, so there probably isn't anything we need to do about this.
![image](https://user-images.githubusercontent.com/8419308/98424095-e6b31a00-204d-11eb-9fb6-99ac09a8b345.png)
![image](https://user-images.githubusercontent.com/8419308/98424414-e2d3c780-204e-11eb-918b-ef99e54aa51e.png)

**Comment by zepumph** [comment]
>It seems like the pickableProperty of disabled buttons remains true. The button gets brighter on hover and the cursor remains a hand. I think this behavior suggests that the disabled button is still interactive, which isn't great UX. It looks like pickable might be problematic and is up for further discussion in #666.

Yes this was done over in https://github.com/phetsims/sun/issues/651, but it was because buttons were being set to pickable: false instead. Talks with @jonathanolson make me feel like this is perhaps the wrong approach. @jonathanolson can you please weigh in on that? Should we move logic back into interactionStateProperties to respect if the buttonModel is disabled? And have the appearance strategies no-op when disabled? Here is a proof-of-concept patch that fixes the hover style, but not the cursor:

```diff
Index: js/buttons/PushButtonInteractionStateProperty.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/buttons/PushButtonInteractionStateProperty.js	(revision 95eadfc4a6fbfd95b5ff3b3f8bbdb6e1eb4ad506)
+++ js/buttons/PushButtonInteractionStateProperty.js	(date 1604708035421)
@@ -22,9 +22,10 @@
     }, options );
 
     super(
-      [ buttonModel.overProperty, buttonModel.looksPressedProperty ],
-      ( over, looksPressed ) => {
-        return over && !looksPressed ? ButtonInteractionState.OVER :
+      [ buttonModel.overProperty, buttonModel.looksPressedProperty, buttonModel.enabledProperty ],
+      ( over, looksPressed, enabled ) => {
+        return !enabled ? ButtonInteractionState.IDLE :
+               over && !looksPressed ? ButtonInteractionState.OVER :
                over && looksPressed ? ButtonInteractionState.PRESSED :
                ButtonInteractionState.IDLE;
       },

```

**Comment by zepumph** [comment]
>The carrot-like arrow used on carousel buttons and the pickers in the Fractions suite gets clipped when the button is disabled. For example:

- [ ] @jonathanolson can you please take a look. I can fix this by removing the `Greyscale.full` filter:

```diff
Index: js/buttons/ButtonNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/buttons/ButtonNode.js	(revision 95eadfc4a6fbfd95b5ff3b3f8bbdb6e1eb4ad506)
+++ js/buttons/ButtonNode.js	(date 1604708444553)
@@ -87,7 +87,7 @@
         background.filters = enabled ? [] : [ CONTRAST_FILTER, BRIGHTNESS_FILTER ];
 
         if ( content ) {
-          content.filters = enabled ? [] : [ Grayscale.FULL ];
+          content.filters = enabled ? [] : [  ];
           content.opacity = enabled ? 1 : SunConstants.DISABLED_OPACITY;
         }
       },

```

**Comment by zepumph** [comment]
>The disabled carousel button in the above screenshot looks odd. From phetsims/balancing-act#97 (comment), it seems like there shouldn't be any background on the disabled button. My guess is that this is a sim-specific issue, but I'm not sure.

- [ ] This is likely because we are hard coding the background color when disabled: ColorConstants.LIGHT_GRAY. Currently this isn't changable. More and more, I'm wondering if we shouldn't have this integrated back into the button appearance strategy. 
 We could also create a new option for it. @jonathanolson what do you think?

**Comment by jonathanolson** [comment]
>> The carrot-like arrow used on carousel buttons and the pickers in the Fractions suite gets clipped when the button is disabled. For example:
>
 > @jonathanolson can you please take a look. I can "fix" this by removing the Greyscale.full filter:

I'll look into this, I'm curious if it's present on different browsers (since it's somewhat unexpected).

> Yes this was done over in #651, but it was because buttons were being set to pickable: false instead. Talks with @jonathanolson make me feel like this is perhaps the wrong approach. @jonathanolson can you please weigh in on that?

Yes, being pickable is a necessary condition to be interactive, but is different. It's more "this component is invisible to any input".

>  This is likely because we are hard coding the background color when disabled: ColorConstants.LIGHT_GRAY. Currently this isn't changable. More and more, I'm wondering if we shouldn't have this integrated back into the button appearance strategy.
We could also create a new option for it. @jonathanolson what do you think?

Having that back in the button appearance strategy (with an option for it) sounds great to me.

**Comment by jonathanolson** [comment]
Added a commit that should work around the cut-off buttons. It unfortunately looks like SVG's filter regions are applied WITHOUT taking strokes into consideration.

**Comment by zepumph** [comment]
To unblock this issue, we can add pickable setting to the enabledAppearanceStrategy in ButtonNode. Then we can work towards a better solution over in https://github.com/phetsims/scenery/issues/1116.

**Comment by zepumph** [comment]
@jonathanolson recommended `node.pickable = enabled? null: false`.

**Comment by zepumph** [comment]
Workaround is implemented above. Unmarking as blocking. All that is left is to deal with the hard coded light grey as the disabled background.

**Comment by zepumph** [comment]
https://github.com/phetsims/sun/issues/658#issuecomment-723357910 was solved in the above commit. Appearance strategies don't have control over it, but it is now an option that can be specified. @jonathanolson how does that sound? Anything else for this issue?

**Comment by jonathanolson** [comment]
> how does that sound? Anything else for this issue?

Sounds good to me.
